Event Handling (#14)
Ada-Developers-Academy/core-unit-3: event-handling/adding-onclick-handling.md

# Adding onclick Handling

## Goals

> Our goals for this lesson is to implement event handling in JavaScript. 

Reviewer: ```suggestion
Our goal for this lesson is to implement event handling in JavaScript. 
```

1. Build an event handler
1. Register the event handler to some HTML element(s)

## Vocabulary

| Vocab         | Definition                                                                                                           | How to Use in a Sentence                                                                                                                                                                                                            |
| ------------- | -------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Event Handler | A function whose responsibility is to perform necessary, required, or desired actions after a certain event is fired | "The function is a click event handler because we expect it to run after the button is clicked," "In order for something to happen after we scroll, we need to make an event handler for what should happen when the user scrolls." |

> 

Reviewer: ## Preparing the Page

Before we look at how to set up our event handler, we need to have an element that can raise the events we want to receive. We should aso set up any other elements that we will modify in response to the event.

We know that we need a button named "Add Crab", so we add the following HTML to our page:

```html
<section>
  <button id="addCrabButton">Add Crab</button>
</section>
```

The `<section>` is optional, but it can be useful to wrap controls in another grouping element to help with styling.

Next, we need to make sure there's place for us to add all those crabs! We can add another section to act as a container for them all.

```html
<section id="crabContainer"></section>
```

And that's it! Moving forward, we'll look at how to listen for events coming from the button, and then add lots and lots of crabs to crab container using an event handler!

## Building Event Handler

In order to build this feature, let's first define an _event handler_ in [our JavaScript code](https://replit.com/@adacore/Adding-onclick-Handling-Demo#scripts/index.js). An **event handler** is a function whose responsibility is to perform necessary, required, or desired actions after a certain event is fired.

Let's make a function named `addCrab`. There's no specific pattern to follow, so we can code it in any style that works. This function:

> - Creates an HTML element that contains a crab
> - Finds the right HTML element to contain the new HTML element
> - Adds the new crab element

Reviewer: I agree with Claire that additional clarification in these points is helpful. I started from Claire's suggestion, but my version is intended to tie into the flow of the other suggestions I'm making.
```suggestion
- Create a `span` element that will contain our new crab
- Find the HTML element that matches the given selector rule (our crab container!), which is where we will add our new `span`
- Set the new crab element (`span`) content to something appropriately "crabby", then add it to our crab container which we just selected
```

### !end-callout

### Example: Registering Handlers to Add Crabs

Let's dive into [our Crab Site example](https://replit.com/@adacore/Adding-onclick-Handling-Demo#scripts/index.js)! We want the `addCrab` handler to run every time we click an "Add Crab" button. Let's first create the HTML element:

```html
<button id="addCrabButton">Add Crab</button>
```

We should also ensure that we have our crab container HTML element somewhere on the page:

```html
<section id="crabContainer"></section>
```

Now, let's register our `addCrab` event handler with our Add Crab button. Observe how we'll have to do an extra step in this process: we'll need to register event handlers as part of handling a different event, "DOMContentLoaded."

```js
const addCrab = () => {
  const newCrab = document.createElement("span");
  const crabContainer = document.querySelector("section#crabContainer");
  newCrab.textContent = "🦀";
  crabContainer.appendChild(newCrab);
};

const registerEventHandlers = () => {
  const crabButton = document.querySelector("#addCrabButton");
  crabButton.addEventListener("click", addCrab);
};

document.addEventListener("DOMContentLoaded", registerEventHandlers);

> ```

Reviewer: ```js
const addCrab = () => {
  const newCrab = document.createElement("span");
  const crabContainer = document.querySelector("#crabContainer");
  newCrab.textContent = "🦀";
  crabContainer.appendChild(newCrab);
};
```

Now we can register our `addCrab` event handler with our "Add Crab" button. Let's examine the following code, which registers our event handler in response to handling a _different_ event: `DOMContentLoaded`.

```js
const registerEventHandlers = () => {
  const crabButton = document.querySelector("#addCrabButton");
  crabButton.addEventListener("click", addCrab);
};

document.addEventListener("DOMContentLoaded", registerEventHandlers);
```

> ```

Author: I end up liking Ansel's suggestion more for one big reason:

When students have to add a chunk of code, and then a "stray" single line after that, I think there's a pattern of accidentally skipping that line.

On that note, later I'll probably add a literal JS comment before `registerEventHandlers` to say `// ... addCrab function`, to pre-emptively disambiguate _where_ (bottom of file) `registerEventHandlers` goes

| `"DOMContentLoaded"`                                | The name of the event that fires when the DOM has loaded completely is "DOMContentLoaded."                                                    |
| `registerEventHandlers`                             | This is the function that should execute when the `document` receives an event named "DOMContentLoaded."                                      |

### !callout-info

## Why Do We Listen for DOMContentLoaded?

We need to ask our document to wait and register the event handlers _after_ the DOM has been completely loaded. This is because the JavaScript file often loads faster than the DOM does! If we don't wait for the DOM to load, then our JS code will not find our `crabButton` in the DOM, and `crabButton` will be `null`. It's important to consider what's on the DOM whenever we search for an element.

<br/>

In other libraries, this problem may already be solved by the library, without us registering a DOMContentLoaded handler.

### !end-callout

### Verifying

Let's check our website now to see if we can add some crabs!

> ![Crab Fan Site, with seven crabs added to the bottom as a result of clicking the "Add Crab" button seven times.](../assets/event-handling_adding-onclick-handling.png)

Reviewer: adds caption
```suggestion
![Crab Fan Site, with seven crabs added to the bottom as a result of clicking the "Add Crab" button seven times.](../assets/event-handling_adding-onclick-handling.png)  
_Fig. The Crab Fan Site, now with extra crabs thanks to JavaScript!_
```

1. A click event that happens on the `addCrab` button at 12:00
1. A click event that happens on the `addCrab` button at 12:01
1. A click event that happens on the `addCrab` button at 12:02, and so on.

By default, our browser will pass in the event object (the object that represents the event) into every event-handler function.

Access to the Event object can give us extra details about the event, and can help us customize our event-handling.

[Let's change our JS code](https://replit.com/@adacore/Adding-onclick-Handling-Demo#scripts/index.js) to anticipate and access the event object that comes into every event-handler. Notice that we are only changing two things:

- The parameters for event-handling functions
- Accessing the events inside the functions in print statements

```js
const addCrab = (event) => {
  console.log("in addCrab:", event);
  // ... other addCrab functionality
};

const registerEventHandlers = (event) => {
  console.log("in registerEventHandlers:", event);
  // ... other registerEventHandler functionality
};

// ...
```

> When we run this, we can see in the console that we get details about the "DOMContentLoaded" event when the page loads. Also, every time we add a crab, we get details about the "click" event (helpfully labeled as `MouseEvent` in the console).

Reviewer: 'also,' feels like it brings everything to a temporary halt. keep it flowing
```suggestion
When we run this, we can see in the console that we get details about the "DOMContentLoaded" event when the page loads. And every time we add a crab, we get details about the "click" event (helpfully labeled as `MouseEvent` in the console).
```

Reviewer: The Event object is worth exploring, but in this context, I worry that it would just be a "here's a bag of stuff!" without a lot of grounding. I'd be a little more inclined to only explicitly introduce the parts of Event they need as they need them, and the rest to the reader's curiosity.

Author: I like the idea of exploring an Event object, but I think that's a tangent that can be explored during this time, during React, or... uh never lol. To Ansel's point, it'd be nice if students had a more concrete objective for exploring it.

Another reason why I'll punt the Event object exploration in text atm is because documentation for Event objects is pretty terrible imo 🙀 aaaaand I'll just sidestep that problem for now by not including it :)
